apps/server/src/modules/tool/uc/external-tool.uc.ts

```diff
@@ -0,0 +1,67 @@
+import { Inject, Injectable, UnprocessableEntityException } from '@nestjs/common';
+import { ICurrentUser, Permission, User } from '@shared/domain';
+import { ExternalToolService } from '@src/modules/tool/service/external-tool.service';
+import { ExternalToolDO, Lti11ToolConfigDO, Oauth2ToolConfigDO } from '@shared/domain/domainobject/external-tool';
+import { AuthorizationService } from '@src/modules/authorization';
+import { OauthProviderService } from '@shared/infra/oauth-provider';
+import { ProviderOauthClient } from '@shared/infra/oauth-provider/dto';
+import { DefaultEncryptionService, IEncryptionService } from '@shared/infra/encryption';
+import { ExternalToolRequestMapper } from '@src/modules/tool/mapper/external-tool-request.mapper';
+
+@Injectable()
+export class ExternalToolUc {
+	constructor(
+		private readonly externalToolService: ExternalToolService,
+		private readonly externalToolMapper: ExternalToolRequestMapper,
+		private readonly authorizationService: AuthorizationService,
+		private readonly oauthProviderService: OauthProviderService,
+		@Inject(DefaultEncryptionService) private readonly oAuthEncryptionService: IEncryptionService
+	) {}
+
+	async createExternalTool(externalToolDO: ExternalToolDO, currentUser: ICurrentUser): Promise<ExternalToolDO> {
+		const user: User = await this.authorizationService.getUserWithPermissions(currentUser.userId);
+		this.authorizationService.checkAllPermissions(user, [Permission.TOOL_CREATE]);
+
+		if (!(await this.externalToolService.isNameUnique(externalToolDO))) {
+			throw new UnprocessableEntityException(`The tool name "${externalToolDO.name}" is already used`);
+		}
+		if (externalToolDO.parameters && this.externalToolService.hasDuplicateAttributes(externalToolDO.parameters)) {
+			throw new UnprocessableEntityException(
+				`The tool: ${externalToolDO.name} contains multiple of the same custom parameters`
+			);
+		}
+		if (externalToolDO.parameters && !this.externalToolService.validateByRegex(externalToolDO.parameters)) {
+			throw new UnprocessableEntityException(
+				`A custom Parameter of the tool: ${externalToolDO.name} has wrong regex attribute.`
+			);
+		}
+
+		if (externalToolDO.config instanceof Lti11ToolConfigDO) {
+			externalToolDO.config.secret = this.oAuthEncryptionService.encrypt(externalToolDO.config.secret);
+		}
+
+		let created: ExternalToolDO;
+		if (externalToolDO.config instanceof Oauth2ToolConfigDO) {
+			if (!(await this.externalToolService.isClientIdUnique(externalToolDO.config))) {
+				throw new UnprocessableEntityException(`The Client Id of the tool: ${externalToolDO.name} is already used`);
+			}
+
+			const oauthClient: ProviderOauthClient = this.externalToolMapper.mapDoToProviderOauthClient(
+				externalToolDO.name,
+				externalToolDO.config
+			);
+			const createdOauthClient = await this.oauthProviderService.createOAuth2Client(oauthClient);
```

Comment:
Nice, i like that we are saving data to hydra now in the server. Now we dont need this orchestration in the shd